File: src/系统/03．技能系统/00．技能模板+函数/02．通用函数/18．单位动画守护.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("lib.扩展函数.封装函数.01．通用工具.02．计时器")
local withTimer = ____require_result_0.withTimer
local stopTimer = ____require_result_0.stopTimer
local ____require_result_1 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_1.registerDeathListener
local GetHandleId = jass.GetHandleId
local GetUnitTypeId = jass.GetUnitTypeId
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
local SetUnitAnimation = jass.SetUnitAnimation
local SetUnitAnimationByIndex = jass.SetUnitAnimationByIndex
local AddUnitAnimationProperties = jass.AddUnitAnimationProperties
local UNIT_ALIVE_LIFE = 0.405
local _____52A8_753B_5B88_62A4_5B9E_4F8B_8868 = {}
local _____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID = 1
local _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = false
local function _____5355_4F4D_5B58_6D3B(unit)
    if unit == nil or unit == 0 then
        return false
    end
    if GetUnitTypeId(unit) == 0 then
        return false
    end
    if IsUnitType(unit, jass.UNIT_TYPE_DEAD) then
        return false
    end
    return GetUnitState(unit, jass.UNIT_STATE_LIFE) > UNIT_ALIVE_LIFE
end
local function _____64AD_653E_5B88_62A4_52A8_753B(_____5B9E_4F8B)
    local _____53C2_6570 = _____5B9E_4F8B["参数"]
    if _____53C2_6570["附加动画属性"] ~= nil and _____53C2_6570["附加动画属性"] ~= "" and type(AddUnitAnimationProperties) == "function" then
        AddUnitAnimationProperties(_____5B9E_4F8B["单位"], _____53C2_6570["附加动画属性"], true)
    end
    if _____53C2_6570["动画编号"] ~= nil then
        SetUnitAnimationByIndex(_____5B9E_4F8B["单位"], _____53C2_6570["动画编号"])
        return
    end
    if _____53C2_6570["动画名"] ~= nil and _____53C2_6570["动画名"] ~= "" then
        SetUnitAnimation(_____5B9E_4F8B["单位"], _____53C2_6570["动画名"])
    end
end
local function _____505C_6B62_5B9E_4F8B(_____5B9E_4F8B)
    if _____5B9E_4F8B["已停止"] then
        return
    end
    _____5B9E_4F8B["已停止"] = true
    stopTimer(_____5B9E_4F8B["计时器"])
    __TS__Delete(_____52A8_753B_5B88_62A4_5B9E_4F8B_8868, _____5B9E_4F8B.ID)
end
local function ____on_52A8_753B_5B88_62A4Tick(_____5B9E_4F8B)
    if _____5B9E_4F8B["已停止"] then
        return
    end
    if not _____5355_4F4D_5B58_6D3B(_____5B9E_4F8B["单位"]) then
        _____505C_6B62_5B9E_4F8B(_____5B9E_4F8B)
        return
    end
    _____64AD_653E_5B88_62A4_52A8_753B(_____5B9E_4F8B)
end
local function ____on_5355_4F4D_6B7B_4EA1(dyingUnit, _killingUnit)
    local _____6B7B_4EA1_5355_4F4DID = GetHandleId(dyingUnit)
    for key in pairs(_____52A8_753B_5B88_62A4_5B9E_4F8B_8868) do
        do
            local _____5B9E_4F8B = _____52A8_753B_5B88_62A4_5B9E_4F8B_8868[key]
            if _____5B9E_4F8B == nil or _____5B9E_4F8B["已停止"] then
                goto __continue16
            end
            if GetHandleId(_____5B9E_4F8B["单位"]) == _____6B7B_4EA1_5355_4F4DID then
                _____505C_6B62_5B9E_4F8B(_____5B9E_4F8B)
            end
        end
        ::__continue16::
    end
end
local function _____786E_4FDD_6B7B_4EA1_76D1_542C()
    if _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C then
        return
    end
    _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = true
    registerDeathListener(____on_5355_4F4D_6B7B_4EA1)
end
____exports["创建单位动画守护"] = function(_____53C2_6570)
    if not _____5355_4F4D_5B58_6D3B(_____53C2_6570["单位"]) then
        return nil
    end
    if _____53C2_6570["动画编号"] == nil and (_____53C2_6570["动画名"] == nil or _____53C2_6570["动画名"] == "") then
        return nil
    end
    if _____53C2_6570["死亡时清理"] ~= false then
        _____786E_4FDD_6B7B_4EA1_76D1_542C()
    end
    local ____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID_2 = _____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID
    _____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID = ____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID_2 + 1
    local ID = ____4E0B_4E00_4E2A_52A8_753B_5B88_62A4ID_2
    local _____95F4_9694_79D2 = _____53C2_6570["间隔秒"] ~= nil and _____53C2_6570["间隔秒"] > 0 and _____53C2_6570["间隔秒"] or 0.5
    local _____5B9E_4F8B = {
        ID = ID,
        ["单位"] = _____53C2_6570["单位"],
        ["计时器"] = nil,
        ["参数"] = _____53C2_6570,
        ["已停止"] = false
    }
    _____52A8_753B_5B88_62A4_5B9E_4F8B_8868[ID] = _____5B9E_4F8B
    if _____53C2_6570["立即播放"] ~= false then
        _____64AD_653E_5B88_62A4_52A8_753B(_____5B9E_4F8B)
    end
    _____5B9E_4F8B["计时器"] = withTimer(
        _____95F4_9694_79D2,
        function() return ____on_52A8_753B_5B88_62A4Tick(_____5B9E_4F8B) end,
        true,
        _____53C2_6570["调试名"] or "单位动画守护"
    )
    return {ID = ID, ["单位"] = _____53C2_6570["单位"]}
end
____exports["停止单位动画守护"] = function(_____53E5_67C4)
    if _____53E5_67C4 == nil then
        return
    end
    local ID = type(_____53E5_67C4) == "number" and _____53E5_67C4 or _____53E5_67C4.ID
    local _____5B9E_4F8B = _____52A8_753B_5B88_62A4_5B9E_4F8B_8868[ID]
    if _____5B9E_4F8B == nil then
        return
    end
    _____505C_6B62_5B9E_4F8B(_____5B9E_4F8B)
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/02．通用函数/18．单位动画守护.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const { withTimer, stopTimer } = require("lib.扩展函数.封装函数.01．通用工具.02．计时器") as {
  withTimer: (this: void, delaySec: number, callback: (this: void) => void, periodic?: boolean, name?: string) => any;
  stopTimer: (this: void, timer: any) => void;
};

const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};

const GetHandleId = jass.GetHandleId as (this: void, handle: any) => number;
const GetUnitTypeId = jass.GetUnitTypeId as (this: void, unit: any) => number;
const GetUnitState = jass.GetUnitState as (this: void, unit: any, state: any) => number;
const IsUnitType = jass.IsUnitType as (this: void, unit: any, unitType: any) => boolean;
const SetUnitAnimation = jass.SetUnitAnimation as (this: void, unit: any, animationName: string) => void;
const SetUnitAnimationByIndex = jass.SetUnitAnimationByIndex as (this: void, unit: any, animationIndex: number) => void;
const AddUnitAnimationProperties = jass.AddUnitAnimationProperties as
  | ((this: void, unit: any, animationProperties: string, add: boolean) => void)
  | undefined;

const UNIT_ALIVE_LIFE = 0.405;

export interface 单位动画守护参数 {
  单位: any;
  动画编号?: number;
  动画名?: string;
  附加动画属性?: string;
  间隔秒?: number;
  立即播放?: boolean;
  死亡时清理?: boolean;
  调试名?: string;
}

export interface 单位动画守护句柄 {
  readonly ID: number;
  readonly 单位: any;
}

interface 单位动画守护实例 {
  ID: number;
  单位: any;
  计时器: any;
  参数: 单位动画守护参数;
  已停止: boolean;
}

const 动画守护实例表: { [id: number]: 单位动画守护实例 | undefined } = {};
let 下一个动画守护ID = 1;
let 已注册死亡监听 = false;

function 单位存活(this: void, unit: any): boolean {
  if (unit == null || unit === 0) return false;
  if (GetUnitTypeId(unit) === 0) return false;
  if (IsUnitType(unit, jass.UNIT_TYPE_DEAD)) return false;
  return GetUnitState(unit, jass.UNIT_STATE_LIFE) > UNIT_ALIVE_LIFE;
}

function 播放守护动画(this: void, 实例: 单位动画守护实例): void {
  const 参数 = 实例.参数;
  if (参数.附加动画属性 != null && 参数.附加动画属性 !== "" && typeof AddUnitAnimationProperties === "function") {
    AddUnitAnimationProperties(实例.单位, 参数.附加动画属性, true);
  }

  if (参数.动画编号 != null) {
    SetUnitAnimationByIndex(实例.单位, 参数.动画编号);
    return;
  }

  if (参数.动画名 != null && 参数.动画名 !== "") {
    SetUnitAnimation(实例.单位, 参数.动画名);
  }
}

function 停止实例(this: void, 实例: 单位动画守护实例): void {
  if (实例.已停止) return;
  实例.已停止 = true;
  stopTimer(实例.计时器);
  delete 动画守护实例表[实例.ID];
}

function on动画守护Tick(this: void, 实例: 单位动画守护实例): void {
  if (实例.已停止) return;
  if (!单位存活(实例.单位)) {
    停止实例(实例);
    return;
  }
  播放守护动画(实例);
}

function on单位死亡(this: void, dyingUnit: any, _killingUnit: any): void {
  const 死亡单位ID = GetHandleId(dyingUnit);
  for (const key in 动画守护实例表) {
    const 实例 = 动画守护实例表[key as any];
    if (实例 == null || 实例.已停止) continue;
    if (GetHandleId(实例.单位) === 死亡单位ID) {
      停止实例(实例);
    }
  }
}

function 确保死亡监听(this: void): void {
  if (已注册死亡监听) return;
  已注册死亡监听 = true;
  registerDeathListener(on单位死亡);
}

export function 创建单位动画守护(参数: 单位动画守护参数): 单位动画守护句柄 | null {
  if (!单位存活(参数.单位)) return null;
  if (参数.动画编号 == null && (参数.动画名 == null || 参数.动画名 === "")) return null;

  if (参数.死亡时清理 !== false) 确保死亡监听();

  const ID = 下一个动画守护ID++;
  const 间隔秒 = 参数.间隔秒 != null && 参数.间隔秒 > 0 ? 参数.间隔秒 : 0.5;
  const 实例: 单位动画守护实例 = {
    ID,
    单位: 参数.单位,
    计时器: null,
    参数,
    已停止: false,
  };

  动画守护实例表[ID] = 实例;
  if (参数.立即播放 !== false) 播放守护动画(实例);
  实例.计时器 = withTimer(间隔秒, () => on动画守护Tick(实例), true, 参数.调试名 ?? "单位动画守护");

  return { ID, 单位: 参数.单位 };
}

export function 停止单位动画守护(句柄: 单位动画守护句柄 | number | null | undefined): void {
  if (句柄 == null) return;
  const ID = typeof 句柄 === "number" ? 句柄 : 句柄.ID;
  const 实例 = 动画守护实例表[ID];
  if (实例 == null) return;
  停止实例(实例);
}

export {};
